### DISCUSSION TITLE
Questions on using the `grpc-bridge-example`.

### DISCUSSION THREAD

**Discussion by liweinan** [resteasy/resteasy-examples#147] [Q&A] [answered]
@ronsigal Hi Ron,

After spending some time on learning the `grpc-bridge` topic, I'm following the README of the `grpc-bridge-example` trying to setup a running example locally. I can successfully build the example and generate the bridge project, and I can setup a WildFly server with gRPC subsystem enabled with the Galleon CLI tool. And the bridge project is deployed correctly:

```bash
02:31:34,972 WARN  [org.jboss.as.dependency.private] (MSC service thread 1-7) WFLYSRV0018: Deployment "deployment.grpcToRest.example.grpc-1.0.0.Final-SNAPSHOT.war" is using a private module ("org.wildfly.extension.grpc") which may be changed or removed in future versions without notice.
02:31:34,972 WARN  [org.jboss.as.dependency.private] (MSC service thread 1-7) WFLYSRV0018: Deployment "deployment.grpcToRest.example.grpc-1.0.0.Final-SNAPSHOT.war" is using a private module ("io.grpc") which may be changed or removed in future versions without notice.
...
02:31:35,167 INFO  [org.wildfly.extension.grpc] (ServerService Thread Pool -- 80) WFLYGRPC0003: gRPC server for grpcToRest.example-1.0.0.Final-SNAPSHOT.war listening on localhost:9555
...
02:31:36,195 INFO  [org.wildfly.extension.undertow] (ServerService Thread Pool -- 81) WFLYUT0021: Registered web context: '/grpcToRest.example.grpc-1.0.0.Final-SNAPSHOT' for server 'default-server'
...
02:31:36,242 INFO  [org.jboss.as.server] (Controller Boot Thread) WFLYSRV0010: Deployed "grpcToRest.example.grpc-1.0.0.Final-SNAPSHOT.war" (runtime-name : "grpcToRest.example.grpc-1.0.0.Final-SNAPSHOT.war")
```

Here is the screenshot of the WildFly output:

<img width="2763" alt="image" src="https://github.com/resteasy/resteasy-examples/assets/201907/9c15504e-25c5-4c51-a20a-034c106b5464">

However if I tried to access the service it returns `404`:

```bash
➤ curl http://localhost:8080/grpcToRest.example.grpc-1.0.0-SNAPSHOT/grpcToJakartaRest/grpcserver/context                                                      02:43:08
...
<html><head><title>Error</title></head><body>404 - Not Found</body></html>
```

Using the test case produced the same result:

<img width="1876" alt="image" src="https://github.com/resteasy/resteasy-examples/assets/201907/4034a364-ad73-4e2a-b250-a8b7e8a04ab7">

I'm not sure what's wrong in my local setup. Could you please give some suggestions on how to troubleshooting the problem? Thanks!

**Comment by jamezp** [comment]
@liweinan This is slightly off topic, but it would be great if we could avoid having to first invoke that context endpoint. If you've got time to experiment feel free to have a look so we can avoid that :)

**Comment by liweinan** [comment]
@jamezp Yes sir!

**Comment by liweinan** [comment]
@ronsigal In addition I have another question: In the generated `grpcToRest.example` project, the `web.xml` is defined like this:

```bash
    <servlet>
        <servlet-name>GreetServlet</servlet-name>
        <servlet-class>
            dev.resteasy.grpc.bridge.runtime.servlet.GrpcHttpServletDispatcher
        </servlet-class>
    </servlet>
```

However I can't find this class anywhere, am I missing anything?

**Reply by ronsigal** [reply]
Heyyy, Weinan!! First of all, I'm so happy that you're playing with this stuff. Thank you!!

1. I don't know what's wrong with the first question. It looks like it should work. I'll go through the example and see if it gives me trouble.
2. dev.resteasy.grpc.bridge.runtime.servlet.GrpcHttpServletDispatcher is in grpc-bridge-runtime: https://github.com/resteasy/resteasy-grpc/blob/main/grpc-bridge-runtime/src/main/java/dev/resteasy/grpc/bridge/runtime/servlet/GrpcHttpServletDispatcher.java

**Reply by ronsigal** [reply]
As for the initial Jakarta REST call, I would like to hide it somehow. One time I tried to make the call from somewhere on the server side, but it didn't work for some reason, and I didn't pursue it at the time. Another option would be to create a client side class that assists in putting together a gRPC call and does the initial call "under the covers". 

The point of the call is to capture a ServletContext for the WAR so that we can reuse the work done inside undertow. Actually, I'm also working on a way to use the same call to capture a HttpServletRequest to use as a delegate for the security methods like HttpServletRequest.isSecure() and HttpServletRequest.login(). I'm on my third attempt to get that done, and I think this is working. I'm about to set up some tests to verify that it works.

**Reply by ronsigal** [reply, marked as answer]
Ok, I've found some problems in grpc-bridge-example, in README.md and GreetingTest. It's embarrassing because I thought I went through everything at least twice. I suppose I read through this one but didn't actually work through the instructions. I should probably work through the rest of the documentation. In any case, I've submitted a pull request (#151) which seems to work. @liweinan, could you try again? Thanks!!

**Reply by ronsigal** [reply]
By the way, there's an annoying little thing in README.md that I never found away around. I wrote
```
"http: //localhost:8080/grpcToRest.example.grpc-1.0.1.Final-SNAPSHOT/grpcToJakartaRest/grpcserver/context
```
(note the space after https: ) because the Eclipse markdown editor didn't render the URL properly. But I've just removed the spaces because the github markdown editor handles it just fine. @liweinan, how does it look to you?

**Comment by liweinan** [comment]
@ronsigal Thanks for the great answers! I'll follow up the instructions and give back my findings here :D

**Comment by liweinan** [comment]
Hi Ron,

With your updated branch I can access the URL now:

```
➤ curl http://localhost:8080/grpcToRest.example.grpc-1.0.1.Final-SNAPSHOT/grpcToJakartaRest/grpcserver/context                                                                                                                                                                              00:21:36

2Got org.greet.Greet_Server@326e9c3a servletContext⏎
```

In addition, I can make the test run:

<img width="964" alt="image" src="https://github.com/resteasy/resteasy-examples/assets/201907/7c821fc3-c1fe-431a-af55-570938f44412">

Here is the workable project in my local environment:

- https://github.com/liweinan/grpcToRest.example


btw I added two comments here:

- https://github.com/resteasy/resteasy-examples/pull/151

Thanks for help!

Guess I can start to write a blog post to report my using experience of the example :D

**Reply by ronsigal** [reply]
@liweinan, this is great news. Thank you!! I've merged the pull request. Would you like to submit another pull request with your two changes?

**Comment by liweinan** [comment]
@ronsigal In addition, while playing with the example, the manual step to edit the TestCase and adding the test scope dependencies creates some obstacles to follow up, maybe the test case and the dependencies could be generated correctly in the archetype generation step(and the testcase can be marked as `@Ignored` and user just need to remove the annotation to run it), wdyt?

**Reply by ronsigal** [reply]
Well, I like the idea of simplifying the process. The problem is that these imports
```
import org.greet.GreetServiceGrpc;
import org.greet.GreetServiceGrpc.GreetServiceBlockingStub;
import org.greet.Greet_proto.GeneralEntityMessage;
import org.greet.Greet_proto.GeneralReturnMessage;
import org.greet.Greet_proto.dev_resteasy_example_grpc_greet___GeneralGreeting;
import org.greet.Greet_proto.dev_resteasy_example_grpc_greet___Greeting;
```
refer to classes that don't exist until you run the archetype, right? As you say, generating it in the archetype would solve that problem, but GreetingTest is specific to this example. Now, the &lt;prefix&gt;_Server.java class **IS** generated by the archetype, so maybe we could have the archetype generate a test based on that resource class, maybe making a call to 
```
    @Path("ready")
    @GET
    public String ready() {
        logger.info("gRPC server ready");
        return "ready";
    }
```
What do you think? Would you like to work on that?

**Comment by liweinan** [comment]
@ronsigal Nice idea! Parsing the resources and generate the test according to the resources is the ultimate solution IMHO.

At the first step, maybe I can try to use a JBang plugin to setup the current test file and the dependencies into correct shape, so the example user and run a single Maven command to make the example run.

If I make any progress on this topic, I'll report it here :D

**Comment by ronsigal** [comment]
> Parsing the resources and generate the test 

@liweinan, I suspect that the test can be hard coded in the archetype. For example, consider
```
        try (
                Client client = ClientBuilder.newClient();
                var response = client.target(
                        "http://localhost:8080/grpc-test/grpcToJakartaRest/grpcserver/context")
                        .request()
                        .get()) {
            final var message = response.getStatus() + ": " + response.readEntity(String.class);
            System.out.println("message: " + message);
            Assert.assertEquals(message, response.getStatus(), 200);
        }
```
in resteasy-grpc/testsuite/grpc-tests/src/test/java/org/jboss/resteasy/test/grpc/AbstractGrpcToJakartaRESTTest.java.

**Comment by ronsigal** [comment]
I don't know much about jBang, but anything you can do to make things simpler has to be a good thing.

**Comment by liweinan** [comment]
@ronsigal okay! I'm going on learning the implementation of the bridge project, and I have a question on this method:

```java
@Path("start")
@GET
public String startGRPC(@Context HttpServletRequest request) throws Exception {
    servletContext = request.getServletContext();
    final Greet_Server server = new Greet_Server();
    new Thread() {
        @SuppressWarnings({"deprecation", "removal"})
        public void run() {
            try {
                if (System.getSecurityManager() == null) {
                    server.start();
                } else {
                    AccessController.doPrivileged((PrivilegedExceptionAction<Void>) () -> {server.start(); return null;});
                }
                logger.info("started gRPC server on port " + PORT);
                server.blockUntilShutdown();
            } catch (Exception e) {
                e.printStackTrace();
            }
        }
    }.start();
    return "Starting gRPC server on port " + PORT;
}
```

If the above method is called multiple times, there will be multiple instances be created and it will cause problem right?

**Reply by ronsigal** [reply]
@liweinan, you are correct sir!

That method is there in case the WAR is deployed to a wildFfy without the wildfly grpc subsystem or to something other than wildfly. It's not necessary in the presence of the grpc subsystem because the grpc subsystem creates its own socket.

**Comment by liweinan** [comment]
@ronsigal Now I understand :D Thanks!

**Reply by ronsigal** [reply]
@liweinan, shouldn't you be asleep? ;-)

**Comment by liweinan** [comment]
@ronsigal In order to go on writing my blog post on the usage of the `grpc-bridge-example`, I need to tag a release of `resteasy-examples`, do you agree?

**Comment by ronsigal** [comment]
Sure, @liweinan, do whatever you need to do.

**Comment by ronsigal** [comment]
I'm trying to remember if there is a reference to grpc-bridge-example somewhere that needs to be updated with a new version ...

**Reply by ronsigal** [reply]
Maybe not. https://resteasy.dev/docs/grpc/ refers to it but not the version.

**Reply by liweinan** [reply]
@ronsigal Thanks for checking!
